### ISSUE TITLE
[BERT/PyTorch] Buffer overflow during gradient accumulation in BERT-L Phase1 pretraining

### ISSUE THREAD

**Issue by raviskolli** [NVIDIA/DeepLearningExamples#522]
BERT/PyTorch

**Describe the bug**
My BERT-L Phase1 pretraining run on 8-node DGX-1 with 32G is failing after 1035 steps due to buffer overflow during gradient accumulation.

**To Reproduce**
I have attached the Dockerfile and Jupyter Notebook that can be used to reproduce this fail.
[Nvbert.zip](https://github.com/NVIDIA/DeepLearningExamples/files/4658296/Nvbert.zip)

**Error Log**
DLL 2020-05-18 05:21:26.672680 - Training Epoch: 0 Training Iteration: 1880  average_loss : 7.9696044921875  step_loss : 7.92578125  learning_rate : 0.0056374535622259435 
DLL 2020-05-18 05:21:38.308501 - Training Epoch: 0 Training Iteration: 1881  average_loss : 7.9833984375  step_loss : 8.140625  learning_rate : 0.005640452207737765 
DLL 2020-05-18 05:21:49.949311 - Training Epoch: 0 Training Iteration: 1882  average_loss : 8.004150390625  step_loss : 8.015625  learning_rate : 0.005643450853249588 
DLL 2020-05-18 05:22:01.907879 - PARAMETER loss_scale : 1024.0 
DLL 2020-05-18 05:22:01.911431 - Training Epoch: 0 Training Iteration: 1882  average_loss : nan  step_loss : 7.8828125  learning_rate : 0.00564644949876141 
DLL 2020-05-18 05:22:13.887836 - PARAMETER loss_scale : 512.0 
DLL 2020-05-18 05:22:13.893405 - Training Epoch: 0 Training Iteration: 1882  average_loss : nan  step_loss : 7.94921875  learning_rate : 0.00564644949876141 
DLL 2020-05-18 05:22:25.522078 - Training Epoch: 0 Training Iteration: 1883  average_loss : 8.006591796875  step_loss : 8.0625  learning_rate : 0.00564644949876141 
DLL 2020-05-18 05:22:37.548128 - PARAMETER loss_scale : 256.0 
DLL 2020-05-18 05:22:37.551992 - Training Epoch: 0 Training Iteration: 1883  average_loss : nan  step_loss : 8.109375  learning_rate : 0.005649448144273231 

**Environment**
* Container version: pytorch:20.01-py3):
* GPUs in the system: 8 node DGX-1 (each node has 8 V100s with 32GB). These are Standard_ND40rs_v2 machines in Azure. 

**Configuration**
* global_size = 32
* local_size = 8
* opt_level              : O2
* cast_model_type        : torch.float16
* patch_torch_functions  : False
* keep_batchnorm_fp32    : True
* master_weights         : True
* loss_scale             : dynamic
* bert_model : bert-large-uncased
* train_batch_size : 2048 
* max_seq_length : 128
* max_predictions_per_seq : 20
* max_steps : 7038
* warmup_proportion : 0.2843
* num_steps_per_checkpoint : 200
* learning_rate : 6e-3
* seed : 42
* fp16 : True
* gradient_accumulation_steps : 32
* allreduce_post_accumulation : True
* allreduce_post_accumulation_fp16 : True
* do_train : True

**Comment by sharathts** [comment]
General rule of thumb: phase1: train_batch_size x num_gpus = 65536
                                     phase2: train_batch_size x num_gpus = 32768
gradient_accumulation_steps = train_batch_size / batch_size_that_fits (batch/gpu mentioned on readme)

with the above parameters - you get 2048 x 64 gpus = 131072, which is twice the expected batch.
set train_batch_size=1024 (65536/64 gpus) and gradient_accumulation_steps=16 (1024/64 batch/gpu)

Also what do the parameters global_size = 32 and local_size = 8 stand for?

**Comment by raviskolli** [comment]
> General rule of thumb: phase1: train_batch_size x num_gpus = 65536
> phase2: train_batch_size x num_gpus = 32768
> gradient_accumulation_steps = train_batch_size / batch_size_that_fits (batch/gpu mentioned on readme)
> 
> with the above parameters - you get 2048 x 64 gpus = 131072, which is twice the expected batch.
> set train_batch_size=1024 (65536/64 gpus) and gradient_accumulation_steps=16 (1024/64 batch/gpu)
> 
> Also what do the parameters global_size = 32 and local_size = 8 stand for?

global_size is world_size
local_size is local_size

You are correct with batch size calculations. My initial post had incorrect values for these parameters. I did set train_batch_size to 1024 and grad_accumulation_steps to 16 for local_batch_size of 64 per GPU in my runs.

**Comment by sharathts** [comment]
Can you try with a different seed?

**Comment by nv-kkudrynski** [comment]
Closing due to no activity.
